File: packages/guards/src/api-response-validate.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ApiResponseValidateOptions {
  action: 'block' | 'warn';
  /** Required status codes or success indicators */
  requireFields?: string[];
  /** Block if error patterns detected */
  blockErrors?: boolean;
}

const ERROR_PATTERNS = [
  /\b(?:error|exception|failure|fatal|critical)\b/i,
  /\b(?:500|502|503|504)\s+(?:internal\s+server|bad\s+gateway|service\s+unavailable)/i,
  /\b(?:timeout|connection\s+refused|ECONNREFUSED|ETIMEDOUT)\b/i,
  /\bstack\s*trace\b/i,
];

export function apiResponseValidate(options: ApiResponseValidateOptions): Guard {
  const blockErrors = options.blockErrors ?? true;
  const required = options.requireFields ?? [];

  return {
    name: 'api-response-validate',
    version: '0.1.0',
    description: 'Validates API response structure and detects error leakage',
    category: 'format',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const violations: string[] = [];
      if (blockErrors) {
        for (const p of ERROR_PATTERNS) { const m = text.match(p); if (m) violations.push(`Error pattern: ${m[0]}`); }
      }
      if (required.length > 0) {
        try {
          const parsed = JSON.parse(text.trim());
          if (typeof parsed === 'object' && parsed !== null) {
            for (const f of required) { if (!(f in parsed)) violations.push(`Missing field: ${f}`); }
          }
        } catch { /* not JSON, skip field check */ }
      }
      const triggered = violations.length > 0;
      return { guardName: 'api-response-validate', passed: !triggered, action: triggered ? options.action : 'allow', latencyMs: Math.round(performance.now() - start), details: triggered ? { violations } : undefined };
    },
  };
}

File: python/open_guardrail/guards/api_response_validate.py
"""Validate API response patterns."""

from __future__ import annotations

import json
import time
from typing import List

from open_guardrail.core import GuardResult

_ERROR_PATTERNS = [
    "500 internal server",
    "timeout",
    "econnrefused",
    "stack trace",
    "traceback (most recent",
    "errno",
    "502 bad gateway",
    "503 service unavailable",
    "connection refused",
]


class _ApiResponseValidate:
    def __init__(
        self,
        *,
        action: str = "warn",
        block_errors: bool = True,
        require_fields: List[str] | None = None,
    ) -> None:
        self.name = "api-response-validate"
        self.action = action
        self.block_errors = block_errors
        self.require_fields = require_fields or []

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        reasons: list[str] = []
        lower = text.lower()

        if self.block_errors:
            found = [
                p for p in _ERROR_PATTERNS if p in lower
            ]
            if found:
                reasons.append("error-patterns")

        if self.require_fields:
            try:
                data = json.loads(text)
                if isinstance(data, dict):
                    missing = [
                        f for f in self.require_fields
                        if f not in data
                    ]
                    if missing:
                        reasons.append("missing-fields")
            except (json.JSONDecodeError, TypeError):
                reasons.append("invalid-json")

        triggered = len(reasons) > 0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="api-response-validate",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=(
                f"API response issues: {', '.join(reasons)}"
                if triggered else None
            ),
            latency_ms=round(elapsed, 2),
            details={
                "reasons": reasons,
            } if triggered else None,
        )


def api_response_validate(
    *,
    action: str = "warn",
    block_errors: bool = True,
    require_fields: List[str] | None = None,
) -> _ApiResponseValidate:
    return _ApiResponseValidate(
        action=action,
        block_errors=block_errors,
        require_fields=require_fields,
    )
